File: py/src/imbrace/resources/conversations.py
from typing import Any, Dict, Optional
from ..http import HttpTransport, AsyncHttpTransport


class ConversationsResource:
    """Conversations domain — Sync.

    @param base - channel-service base URL (gateway/channel-service)
    """

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    def list(self, type: Optional[str] = None, q: Optional[str] = None,
             limit: Optional[int] = None, skip: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if type:
            params["type"] = type
        if q:
            params["q"] = q
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip

        # Use v1/conversations if no specific v2 params are used, to match tests
        path = f"{self._v1}/conversations" if not params else f"{self._v2}/team_conversations"
        return self._http.request("GET", path, params=params).json()

    def get(self, conv_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._v1}/team_conversations/{conv_id}").json()

    def get_views_count(self, type: Optional[str] = None, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if type:
            params["type"] = type
        if q:
            params["q"] = q
        return self._http.request("GET", f"{self._v2}/team_conversations/_views_count",
                                  params=params).json()

    def search(self, business_unit_id: str, q: str,
               limit: Optional[int] = None, skip: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {
            "business_unit_id": business_unit_id,
            "type": "text",
            "q": q,
        }
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip
        return self._http.request("GET", f"{self._v1}/team_conversations/_search", params=params).json()

    def create(self) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v1}/conversations").json()

    def join(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v1}/team_conversations/_join", json=body).json()

    def leave(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v1}/team_conversations/_leave", json=body).json()

    def update_status(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._v1}/team_conversations/_update_status", json=body).json()


class AsyncConversationsResource:
    """Conversations domain — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip("/")

    @property
    def _v1(self) -> str:
        return f"{self._base}/v1"

    @property
    def _v2(self) -> str:
        return f"{self._base}/v2"

    async def list(self, type: Optional[str] = None, q: Optional[str] = None,
                   limit: Optional[int] = None, skip: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if type:
            params["type"] = type
        if q:
            params["q"] = q
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip

        path = f"{self._v1}/conversations" if not params else f"{self._v2}/team_conversations"
        res = await self._http.request("GET", path, params=params)
        return res.json()

    async def get(self, conv_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._v1}/team_conversations/{conv_id}")
        return res.json()

    async def get_views_count(self, type: Optional[str] = None, q: Optional[str] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {}
        if type:
            params["type"] = type
        if q:
            params["q"] = q
        res = await self._http.request("GET", f"{self._v2}/team_conversations/_views_count",
                                       params=params)
        return res.json()

    async def search(self, business_unit_id: str, q: str,
                     limit: Optional[int] = None, skip: Optional[int] = None) -> Dict[str, Any]:
        params: Dict[str, Any] = {
            "business_unit_id": business_unit_id,
            "type": "text",
            "q": q,
        }
        if limit is not None:
            params["limit"] = limit
        if skip is not None:
            params["skip"] = skip
        res = await self._http.request("GET", f"{self._v1}/team_conversations/_search", params=params)
        return res.json()

    async def create(self) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._v1}/conversations")
        return res.json()

File: ts/src/resources/conversations.ts
import { HttpTransport } from "../http.js"

export class ConversationsResource {
  /**
   * @param base - channel-service base URL (gateway/channel-service)
   */
  constructor(private readonly http: HttpTransport, private readonly base: string) {}

  private get v1() { return `${this.base}/v1` }
  private get v2() { return `${this.base}/v2` }

  // ─── Team Conversations ──────────────────────────────────────────────────────

  async list(params?: {
    type?: string
    q?: string
    limit?: number
    skip?: number
    sort?: string
  }) {
    const url = new URL(`${this.v2}/team_conversations`)
    if (params?.type)   url.searchParams.set("type",  params.type)
    if (params?.q)      url.searchParams.set("q",     params.q)
    if (params?.limit)  url.searchParams.set("limit", String(params.limit))
    if (params?.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    if (params?.sort)   url.searchParams.set("sort",  params.sort)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async get(convId: string) {
    return this.http.getFetch()(`${this.v1}/team_conversations/${convId}`, { method: "GET" }).then(r => r.json())
  }

  async getByConversationId(conversationId: string) {
    const url = new URL(`${this.v1}/team_conversations`)
    url.searchParams.set("type", "conversation_id")
    url.searchParams.set("q",    conversationId)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async search(params: { businessUnitId: string; q: string; limit?: number; skip?: number }) {
    const url = new URL(`${this.v1}/team_conversations/_search`)
    url.searchParams.set("business_unit_id", params.businessUnitId)
    url.searchParams.set("type",             "text")
    url.searchParams.set("q",               params.q)
    if (params.limit) url.searchParams.set("limit", String(params.limit))
    if (params.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getViewsCount(params?: { type?: string; q?: string }) {
    const url = new URL(`${this.v2}/team_conversations/_views_count`)
    if (params?.type) url.searchParams.set("type", params.type)
    if (params?.q)    url.searchParams.set("q",    params.q)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getOutstanding(params: { businessUnitId: string; limit?: number; skip?: number }) {
    const url = new URL(`${this.v1}/team_conversations/_outstanding`)
    url.searchParams.set("type", "business_unit_id")
    url.searchParams.set("q",    params.businessUnitId)
    if (params.limit) url.searchParams.set("limit", String(params.limit))
    if (params.skip !== undefined) url.searchParams.set("skip", String(params.skip))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async join(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_join`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async leave(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_leave`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateStatus(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_update_status`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async updateName(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_update_name`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async initVideoCall(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_init_jaas_conference`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async assignTeamMember(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/assign_team_member`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async removeTeamMember(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/remove_team_member`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async getInvitableUsers(teamConvId: string) {
    return this.http.getFetch()(`${this.v1}/team_conversations/${teamConvId}/users`, { method: "GET" }).then(r => r.json())
  }

  // ─── Single Conversation ─────────────────────────────────────────────────────

  async getConversation(conversationId: string) {
    return this.http.getFetch()(`${this.v1}/conversations/${conversationId}`, { method: "GET" }).then(r => r.json())
  }

  // ─── Create standalone conversation ─────────────────────────────────────────

  async create(body?: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/conversations`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body ?? {}),
    }).then(r => r.json())
  }

  async joinRequest(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.v1}/team_conversations/_join_request`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }
}
